File: LeetCode/src/math/PalindromeNumber.java
package math;

public class PalindromeNumber {
    public boolean isPalindrome(int x) {

        // negative number and number end with 0
        if (x < 0 || (x != 0 && x % 10 == 0)) {
            return false;
        }

        int reversed = 0;
        int origin = x;

        while (origin != 0) {
            int num = origin % 10;
            reversed = reversed * 10 + num;
            origin /= 10;
        }

        return reversed == x;
    }

    public static void main(String[] args) {
        PalindromeNumber solution = new PalindromeNumber();
        int x1 = 121;
        int x2 = -121;
        System.out.println(solution.isPalindrome(x1));// Output: true
        System.out.println(solution.isPalindrome(x2));// Output: false
    }
}

File: LeetCode_Python/math/PalindromeNumber.py
class PalindromeNumber:
    def isPalindrome(self, x: int) -> bool:
        if x < 0:
            return False

        cur = x
        rev = 0
        while cur:
            rev = rev * 10 + cur % 10
            cur //= 10

        return rev == x


# Time Complexity: O(log10(n))
# Space Complexity: O(1)

if __name__ == "__main__":
    solution = PalindromeNumber()

    x1 = 121
    print("Expected Output: True")
    print("Actual Output:", solution.isPalindrome(x1))

    x2 = -121
    print("Expected Output: False")
    print("Actual Output:", solution.isPalindrome(x2))

    x3 = 10
    print("Expected Output: False")
    print("Actual Output:", solution.isPalindrome(x3))
